pkg/errors/broadcasting.go

```diff
@@ -0,0 +1,213 @@
+type BroadcastOperation string
+
+const (
+	// BackgroundBroadcast represents a background broadcasting operation.
+	BackgroundBroadcast BroadcastOperation = "backgroundBroadcast"
+
+	// ImmediateBroadcast represents an immediate broadcasting operation.
+	ImmediateBroadcast BroadcastOperation = "immediateBroadcast"
+
+	// DelayedBroadcast represents a delayed broadcasting operation.
+	DelayedBroadcast BroadcastOperation = "delayedBroadcast"
+
+	// CreateAction represents an action to create a new wallet action.
+	CreateAction BroadcastOperation = "createAction"
+
+	// ProcessAction represents an action to process an existing wallet action.
+	ProcessAction BroadcastOperation = "processAction"
+)
+
+// BroadcastingError represents an error that occurred during transaction broadcasting
+type BroadcastingError struct {
+	Err             error
+	TxID            string
+	Reference       string
+	SendWithResults []wdk.SendWithResult
+	ReviewResults   []wdk.ReviewActionResult
+	ServiceErrors   map[string]error
+	Operation       BroadcastOperation
+	Tx              []byte
+	NoSendChange    []wdk.OutPoint
+}
+
+// Error implements the error interface for BroadcastingError
+func (e *BroadcastingError) Error() string {
+	if e.Err == nil {
+		return fmt.Sprintf("broadcasting failed during %s for txID %s (reference: %s)",
+			e.Operation, e.TxID, e.Reference)
+	}
+	return fmt.Sprintf("broadcasting failed during %s for txID %s (reference: %s): %v",
+		e.Operation, e.TxID, e.Reference, e.Err)
+}
+
+// Unwrap implements the errors.Unwrap interface for BroadcastingError
+func (e *BroadcastingError) Unwrap() error {
+	return e.Err
+}
+
+// Is implements the errors.Is interface for BroadcastingError
+func (e *BroadcastingError) Is(target error) bool {
+	if target == nil {
+		return false
+	}
+
+	if _, ok := target.(*BroadcastingError); ok {
+		return true
+	}
+
+	if e.Err != nil {
+		return errors.Is(e.Err, target)
+	}
+
+	return false
+}
+
+// EnhanceWithCreateActionContext enhances the error with context specific to a CreateAction operation
+func EnhanceWithCreateActionContext(
+	err error,
+	txID, reference string,
+	tx []byte,
+	noSendChange []wdk.OutPoint,
+	processResult *wdk.ProcessActionResult,
+) error {
+	var broadcastErr *BroadcastingError
+	if errors.As(err, &broadcastErr) {
+		return broadcastErr.
+			WithOperation(CreateAction).
+			WithReference(reference).
+			WithTxIDIfEmpty(txID).
+			WithProcessActionContext(processResult, txID, reference).
+			WithTransactionData(tx, noSendChange)
+	}
+
+	return NewBroadcastingError(err, CreateAction).
+		WithTxID(txID).
+		WithReference(reference).
+		WithTransactionData(tx, noSendChange).
+		WithProcessActionContext(processResult, txID, reference)
+}
+
+// WithContext adds context values from processAction result
+func (e *BroadcastingError) WithContext(processResult *wdk.ProcessActionResult, txID, reference string) *BroadcastingError {
```

Comment:
IMHO this whole concept of WithXyz is adding unneeded complexity to the whole thing.

Actually you could fill the structure inline when returning this error.
Or if you dedicated methods to initialize it, then there should be a constructor for given broadcasting state.